pkg/client/typed/scale/scale.go

```diff
@@ -0,0 +1,218 @@
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type Client struct {
+	dc *dynamic.Client
+}
+
+type ScaleInterface interface {
+	Replicas() (int, error)
+	SetReplicas(replicas int)
+	Selector() (string, error)
+}
+
+func NewClient(conf *restclient.Config) (*Client, error) {
+	// Make a copy, do not modify the passed config.
+	// TODO(madhusudancs): Should we just receive the conf argument by value? Receiving this as a
+	// pointer now just to be consistent with other clients.
+	conf2 := *conf
+	if conf2.APIPath == "" {
+		conf2.APIPath = "/apis"
+	}
+
+	dc, err := dynamic.NewClient(&conf2)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		dc: dc,
+	}, nil
+}
+
+// NewForConfigOrDie creates a new dynamic client for the given config and
+// panics if there is an error in the config.
+func NewForConfigOrDie(conf *restclient.Config) *Client {
+	cl, err := NewClient(conf)
+	if err != nil {
+		panic(err)
+	}
+	return cl
+}
+
+// TODO(madhusudancs): Implement this as a codec decoder.
+func (c *Client) decode(res *runtime.Unstructured) (ScaleInterface, error) {
+	if res.TypeMeta.Kind != "Scale" {
+		return nil, fmt.Errorf("object should be of Kind `Scale`, got `%s`", res.TypeMeta.Kind)
+	}
+	switch res.TypeMeta.APIVersion {
+	case "extensions/v1beta1":
+		return &extScale{scale{c, res}}, nil
+	case "autoscaling/v1":
+		return &asScale{scale{c, res}}, nil
+	default:
+		return nil, fmt.Errorf("unknown APIVersion `%s` for `Scale`", res.TypeMeta.APIVersion)
+	}
+}
+
+func (c *Client) Get(scaleRef extensions.SubresourceReference, namespace string) (ScaleInterface, error) {
+	gv, err := unversioned.ParseGroupVersion(scaleRef.APIVersion)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing group version from scale reference: %v", err)
+	}
+	gvr, _ := meta.KindToResource(gv.WithKind(scaleRef.Kind))
+
+	resource := &unversioned.APIResource{
+		Name:       gvr.Resource,
+		Namespaced: true,
+		Kind:       scaleRef.Kind,
+	}
+	res, err := c.dc.Subresource(resource, scaleRef.Subresource, namespace).Get(scaleRef.Name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch resource %s(%s): %v", scaleRef.Name, scaleRef.Kind, err)
+	}
+	return c.decode(res)
+}
+
+func (c *Client) Update(scaleRef extensions.SubresourceReference, namespace string, obj ScaleInterface) (ScaleInterface, error) {
+	gv, err := unversioned.ParseGroupVersion(scaleRef.APIVersion)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing group version from scale reference: %v", err)
+	}
+	gvr, _ := meta.KindToResource(gv.WithKind(scaleRef.Kind))
+
+	resource := &unversioned.APIResource{
+		Name:       gvr.Resource,
+		Namespaced: true,
+		Kind:       scaleRef.Kind,
+	}
+	var res *runtime.Unstructured
+	switch t := obj.(type) {
```

Comment:
> @deads2k if you're asking why the client looks at the type of the object instead of having a hardcoded map of URL to type, well, why not? This is actually easier.

Let's say you want to issue an update using this client.  Ordinarily, you'd just new up an object an call update, like this: `client.Update(scaleRef, namespace, &internalapi.Scale{replicas: 12}`.

With this structure, I have to know the type I'm going to use ahead of time.  How do I do that?  Well the current calling code assumes that it has `get` access and does a read first, then an `update`.  Otherwise, I'd look it up in discovery.

Let's say I want to extend this to the next logical location: the `kubectl scale` command.  Well, it turns out that I can't do that well because this API wasn't built to hide the ugly of the different `Scale` types from me.  I'll need to use the discover doc to figure it out.  Meaning that the core work has to be done either way and we're just deciding whether we're going to fan that work out to every caller or if we're going to hide it.  I'd rather hide it.